core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala

```diff
@@ -200,20 +187,52 @@ class CassandraJournal(cfg: Config) extends AsyncWriteJournal with CassandraReco
               case write :: remainder => writeMessages(write).flatMap(result => rec(remainder, result :: acc))
               case Nil                => Future.successful(acc.reverse)
             }
+
           rec(groups, Nil)
         }
 
-      result.onComplete { _ =>
+      val tagWrites = result.map({ _ =>
+        val messagesByTag: Map[String, Seq[Serialized]] = serialized
+          .flatMap(_.payload)
+          .flatMap(s => s.tags.map((_, s)))
+          .groupBy(_._1).mapValues(_.map(_._2))
+
+        messagesByTag.map {
+          case (tag, writes) =>
+            TagWrites(tag, writes.toVector)
+        }.toVector
+      })
+
+      tagWrites pipeTo self
```

Comment:
what if the `result` is a failed Future? then we don't know if the write to messages table was successful or not so we can't write the tag. The persistent actor will be stopped and a new recovery will follow.

Note that `pipeTo` will send `akka.actor.Status.Failure`, which is not handled by this actor.

Reply:
Apart from the fact that we'd get an unhandled msg I think this should work as expected. As it may or may not be in the `message` table and definitely won't be in the `tag_view` and `tag_progress` tables so it'll get replayed to the tag writer on recovery

Reply:
👍